### ISSUE TITLE
`SystemStackError` after upgrade to bundler 1.16 

### ISSUE THREAD

**Issue by nelhage-stripe** [rubygems/rubygems#3365]
# Error Report

## Questions

Please fill out answers to these questions, it'll help us figure out
why things are going wrong.

- **What did you do?**

  I ran the command `/usr/lib/rbenv/versions/2.4/bin/bundle install`

- **What did you expect to happen?**

`bundle install` would complete successfully.

- **What happened instead?**

We (Stripe) maintain a large Ruby application internally. After upgrading to bundler 1.16.1 from 1.15.4 we see reproducible `SystemStackError`s from infinite recursion in our CI environment (but only in CI; I have not been able to reproduce locally yet, so it depends on some combination of the environment and our CI settings).

- **Have you tried any solutions posted on similar issues in our issue tracker, stack overflow, or google?**

I have searched but haven't found anything obviously similar.

I'm able to (e.g.) bisect or capture a CI build and drop in a `pry`, but because the issue only reproduces in CI so far doing so is slightly involved, so I've opened an issue first in the hopes that you'll be able to point us at productive things to try.

I've redacted our `Gemfile` and `Gemfile.lock` since they are large and contain various internal references; I'd be happy to provide them to the `bundler` team privately if useful.

- **Have you read our issues document, https://github.com/bundler/bundler/blob/master/doc/contributing/ISSUES.md?**

 Yes.

## Backtrace

```
SystemStackError: stack level too deep
  /usr/lib/rbenv/versions/2.4/lib/ruby/2.4.0/rubygems/stub_specification.rb:172:in `extensions'
  /usr/lib/rbenv/versions/2.4/lib/ruby/2.4.0/rubygems/stub_specification.rb:148:in `missing_extensions?'
  [ following two lines repeat many times ]
  /usr/lib/rbenv/versions/2.4/lib/ruby/2.4.0/rubygems/stub_specification.rb:151:in `missing_extensions?'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/stub_specification.rb:38:in `missing_extensions?'
  /usr/lib/rbenv/versions/2.4/lib/ruby/2.4.0/rubygems/stub_specification.rb:151:in `missing_extensions?'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/stub_specification.rb:38:in `missing_extensions?'
  /usr/lib/rbenv/versions/2.4/lib/ruby/2.4.0/rubygems/stub_specification.rb:151:in `missing_extensions?'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/stub_specification.rb:38:in `missing_extensions?'
  /usr/lib/rbenv/versions/2.4/lib/ruby/2.4.0/rubygems/stub_specification.rb:151:in `missing_extensions?'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/stub_specification.rb:38:in `missing_extensions?'
  /usr/lib/rbenv/versions/2.4/lib/ruby/2.4.0/rubygems/stub_specification.rb:151:in `missing_extensions?'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/stub_specification.rb:38:in `missing_extensions?'
  /usr/lib/rbenv/versions/2.4/lib/ruby/2.4.0/rubygems/stub_specification.rb:151:in `missing_extensions?'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/stub_specification.rb:38:in `missing_extensions?'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/rubygems_integration.rb:76:in `spec_missing_extensions?'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/source/rubygems.rb:347:in `block (2 levels) in installed_specs'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/source/rubygems.rb:344:in `reverse_each'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/source/rubygems.rb:344:in `block in installed_specs'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/index.rb:11:in `build'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/source/rubygems.rb:343:in `installed_specs'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/source/rubygems.rb:91:in `specs'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/lazy_specification.rb:76:in `__materialize__'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/spec_set.rb:85:in `block in materialize'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/spec_set.rb:82:in `map!'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/spec_set.rb:82:in `materialize'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/definition.rb:203:in `missing_specs'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/definition.rb:208:in `missing_specs?'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/installer.rb:276:in `resolve_if_needed'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/installer.rb:83:in `block in run'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/process_lock.rb:12:in `block in lock'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/process_lock.rb:9:in `open'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/process_lock.rb:9:in `lock'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/installer.rb:72:in `run'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/installer.rb:25:in `install'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/cli/install.rb:65:in `run'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/cli.rb:224:in `block in install'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/settings.rb:136:in `temporary'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/cli.rb:223:in `install'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/cli.rb:27:in `dispatch'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/cli.rb:18:in `start'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/exe/bundle:30:in `block in <top (required)>'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/friendly_errors.rb:122:in `with_friendly_errors'
  /usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/exe/bundle:22:in `<top (required)>'
  /usr/lib/rbenv/versions/2.4/bin/bundle:23:in `load'
  /usr/lib/rbenv/versions/2.4/bin/bundle:23:in `<main>'
```

## Environment

```
Bundler       1.16.1
  Platforms   ruby, x86_64-linux
Ruby          2.4.3p205 (2017-12-14 revision 61247) [x86_64-linux]
  Full Path   /usr/lib/rbenv/versions/2.4/bin/ruby
  Config Dir  /etc
RubyGems      2.6.14
  Gem Home    /src/vendor/bundle/ruby/2.4.0
  Gem Path    
  User Path   /root/.gem/ruby/2.4.0
  Bin Dir     /src/vendor/bundle/ruby/2.4.0/bin
OpenSSL       
  Compiled    OpenSSL 1.0.1f 6 Jan 2014
  Loaded      OpenSSL 1.0.1f 6 Jan 2014
  Cert File   /usr/lib/ssl/cert.pem
  Cert Dir    /usr/lib/ssl/certs
Tools         
  Git         2.17.0
  RVM         not installed
  rbenv       rbenv 0.4.0

  chruby      not installed
```

## Bundler Build Metadata

```
Built At          2017-12-21
Git SHA           0034ef341
Released Version  true
```

## Bundler settings

```
frozen
  Set for your local app (/src/.bundle/config): true
path
  Set for your local app (/src/.bundle/config): "vendor/bundle"
jobs
  Set for your local app (/src/.bundle/config): "16"
clean
  Set for your local app (/src/.bundle/config): "true"
disable_shared_gems
  Set for your local app (/src/.bundle/config): true
without
  Set for your local app (/src/.bundle/config): [:monster, :ci_ignore, :test_ui, :no_nodejs, :no_zookeeper]
build.nokogiri
  Set via BUNDLE_BUILD__NOKOGIRI: "--use-system-libraries"
gemfile
  Set via BUNDLE_GEMFILE: "/src/Gemfile"
bin_path
  Set via BUNDLE_BIN_PATH: "/usr/lib/rbenv/versions/2.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/exe/bundle"
```

--- TEMPLATE END ----------------------------------------------------------------

**Comment by anna328p** [comment]
Getting a very similar issue with Bundler 2.1.4: https://hastebin.com/qoyorikeni

**Comment by deivid-rodriguez** [comment]
The fix was released with bundler 2.2, so you need to upgrade to get it.

**Comment by anna328p** [comment]
> The fix was released with bundler 2.2, so you need to upgrade to get it.

Thanks. Looks like I'll have to poke the nixpkgs bundler maintainers to update it.
